### DISCUSSION TITLE
Implement own document source

### DISCUSSION THREAD

**Discussion by Aed3r** [carrot2/carrot2#118] [Q&A] [answered]
Hello,
I started looking into making my own document source using the Reddit API. So far I found the IDocumentSource documentation that loosely explains how to do it, but I can't quite get it to work using just that. 

I tried looking for implementations of existing document sources (Bing, PubMed...) but can't seem to find them.

Where could I find them to replicate how they work? What do I need to watch out for?

Kind regards

**Comment by stanislawosinski** [comment, marked as answer]
Hi Gustav,

In the 4.x line of Carrot2, we moved the document sourcing part (the IDocumentSource known from the 3.x line) out of the core Java API -- the API is now concerned only with clustering, leaving the task of fetching documents to the client calling the API. 

Depending on your use case, I see two options:

1. If your goal is to have Reddit content clustered and exposed in your own user interface, then put the Reddit content fetching in your code and submit that content for Carrot2 clustering.
2. If your goal is to fetch and cluster Reddit content in one of the end-user Carrot2 apps (search results clustering, Carrot2 Workbench), then you can fork Carrot2 and add Reddit source to the code of the end-user apps. Follow the imports from [sources.js](https://github.com/carrot2/carrot2/blob/master/dcs/contexts/frontend/app/src/carrot2/sources.js) for code examples. Alternatively, for quick experiments you can prepare Reddit samples in JSON or CSV format and [cluster them inside Workbench](https://carrot2.github.io/release/4.2.1/doc/getting-started/#excel-open-office-or-csv-spreadsheet). Or, you can index Reddit in Elasticsearch or Solr and [use Workbench to cluster data from the index](https://carrot2.github.io/release/4.2.1/doc/getting-started/#solr-or-elasticsearch).

**Reply by Aed3r** [reply]
Seems I stumbled on an older version, that explains my difficulties. The first use case seems most adequate for what I need, I'll look into that. Thanks a lot!
